Items can now be autopicked-up to nested containers.

moxian opened the pull request:
#### Summary

SUMMARY: Bugfixes "Fixes autopickup into nested containers"

#### Purpose of change

Currently, only pocket autopickup settings on worn items are respected. This is due to
1) that we explicitly ignore nested containers when searching for best pockets UNLESS the top-level container is rigid. This logic almost makes sense (why would you *want* to put items into non-rigid nested containers? You won't save space that way!), but only as an optimization, and only sometimes. With the default pocket settings on everything [and with this fix], you won't be needlessly nesting things anyway.
2) a typo that was comparing a pocket with itself to see if it's better than *another* pocket. That comparison always returned false, i.e. "nah, this pocket is good enough, let's not use that other with priority 1234123"

#### Describe the solution

- remove (uncommented) rigidity check
- fix the typo

#### Testing

Manual.
Spawn a wallet and some money, set wallet's whitelist to money and priority to 1, put wallet in a backpack, pick up money, see it ending up in the wallet, not in a backup.
Set wallet priority back to 0, pick up some more money, see they end up in a wallet again, so as to stack. See that cash cards end up in a backpack.

#### Additional context

I believe this is *the* main reason for the "pockets are just terribly broken and don't work at all!!11!!" sentiment I've seen when trying to google and understand how does the priority work.

Qrox commented on src/item_contents.cpp:
```diff
@@ -507,7 +504,8 @@ std::pair<item_location, item_pocket *> item_contents::best_pocket( const item &
             for( item *contained : all_items_top( item_pocket::pocket_type::CONTAINER ) ) {
                 std::pair<item_location, item_pocket *> internal_pocket =
                     contained->contents.best_pocket( it, parent, true );
```
```suggestion
                    contained->contents.best_pocket( it, parent );
```
Please ensure all places where `item_contents::best_pocket` is used are updated.

moxian replied:
I have updated that but somehow not committed.. Weird.
Thanks for the diligence!

moxian commented:
I did see the test failures in my local directory, but assumed those are just flukes or have something to do with my machine. But seeing them on CI as well is really really concerning.
I'll take a look.

moxian commented:
Quick status update.

I've looked into the test failures. The test failure is intended. It complains that it expects military medic to start with `bandages (9)` and `bandages (3)` in separate containers, but this PR combines them into a single `bandages (12)`.
I have not yet looked into how to fix this.

However during development i did encounter a bug very similar to #46904 (i.e. container and the item disappearing on pickup). I *think* that was because during one of the iterations I accidentally swapped arguments, and was inserting the container into the thing, not the other way around.
I think i've fixed the bug and i couldn't reproduce it since. So the code "works", but that doesn't give me enough confidence.

I think at the least we'll need a `debugmsg` when the item disappears that way, so that we know (from bug reports) if i messed up, and introduced/uncovered a bug that was so far hidden.

mqrause commented:
Are you still working on this? I wanted to fix mainly your point 2, but this will create an issue. `item_pocket::better_pocket` will ultimately compare the time it takes to access the contents. However this will not take into account the time to access the parent pockets, so it's possible items will be put into some nested containers for no good reason.

moxian commented:
Oh, sorry, I'm not working on this. I should have closed this when i realized, but i forgot.
Apologies, i'll do this now.
